### DISCUSSION TITLE
Why does Salvo `RateLimiter` call `ctrl.skip_rest()`?

### DISCUSSION THREAD

**Discussion by TheAwiteb** [salvo-rs/salvo#852] [Q&A] [answered]
Hi,

I decided to delve deeper into how the middleware works in Salvo. I noticed
that if the middleware modifies the response, Salvo will not allow the response
to continue to the endpoint. Instead, it returns the response generated by
the middleware, as [shown here][1]. If the response doesn't have a body, it
retrieves one from the catcher, as [shown here][2].

Given this behavior, why does `RateLimiter` call `ctrl.skip_rest()`, as [shown here][3]?
It seems unnecessary since the function should simply exit via `return`.

Am I correct, or am I missing something?

[1]: https://github.com/salvo-rs/salvo/blob/6d4dca5a25e7d5305060a7a2e86fa9a0e1b0aa44/crates/core/src/routing/mod.rs#L558-L559
[2]: https://github.com/salvo-rs/salvo/blob/6d4dca5a25e7d5305060a7a2e86fa9a0e1b0aa44/crates/core/src/service.rs#L260-L266
[3]: https://github.com/salvo-rs/salvo/blob/6d4dca5a25e7d5305060a7a2e86fa9a0e1b0aa44/crates/rate-limiter/src/lib.rs#L164-L225

**Comment by chrislearn** [comment, marked as answer]
After each handler is called, salvo checks the response. If its status code is an error, the subsequent handlers execution is terminated.

In `RateLimiter`, you can indeed not `call ctrl.skip_rest()`, but I think it is more logical to call it.

**Reply by TheAwiteb** [reply]
Thank you for the quick answer, so my understanding is correct.
